### PR TITLE
Add Query Users API

### PR DESCRIPTION
**Pull request by jfreden**
This PR adds a new Query Users API:
- `POST /_security/_query/user` 
- `GET /_security/_query/user` 

It follows a similar convention to the [Query API key API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html). 

Sample request 
```json
POST /_security/_query/user
{
  "query": { 
    "bool": {
      "must": [
        {
          "wildcard": {
            "username": "org-user*"
          }
        }
      ],
      "filter": [
        {
          "prefix": {
            "roles": {
              "value": "super"
            }
          }
        }
      ],
      "must_not": [
        {
          "term": {
            "enabled": "false"
          }
        }
      ]
    }
  },
  "from": 0,
  "size": 1000,
  "sort": [ 
    "username"
  ],
  "search_after": [
    "org-user-2048"
  ]
}
```

The query supports a subset of query types, including `match_all`, `bool`, `term`, `terms`, `prefix`, `wildcard`, and `exists`.

The supported fields are:
```
username
roles
enabled
full_name
email
```

The query also supports pagination-related fields, analogous to the generic [Search API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html).

The response format is similar to that of the [Get Users API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-user.html) with one key difference: the retrieved users are returned as a list of payloads, instead of as a nested object keyed by username. This is more consistent with the format for a paginated response. 
```json
{
  "total": 3,
  "count": 3,
  "users": [
    { 
    "username": "org-user-2049",
    "roles": [
      "superuser"
    ],
    "full_name": "Jack Nicholson",
    "email": "jacknich@example.com",
    "metadata": { "intelligence" : 7 },
    "enabled": true
  }
  ]
}
```

**Excluded**
- Documentation will be added in https://github.com/elastic/elasticsearch/pull/104255
- `with_profile_id` will be added in a follow up PR ES-7599

### CODE REVIEW

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/privilege/PrivilegeTests.java [hunk 1]
```diff
@@ -339,16 +341,11 @@ public void testManageUserProfilePrivilege() {
             "cluster:admin/xpack/security/role/get",
             "cluster:admin/xpack/security/role/delete"
         );
-        verifyClusterActionDenied(
```
Comment (jfreden) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/privilege/PrivilegeTests.java:342]: This was duplicated so did a drive-by remove.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexFieldNameTranslator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public class SecurityIndexFieldNameTranslator {
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexFieldNameTranslator.java:14]: I added this class to support prefix-field name matching, since that's used in the query apikey api. Now I only use this for the query user api, where we only have exact field name mappings and they all happen to match exactly what's in the security index, so translations are not required. For future-proofing it could be good to keep this, but I also understand that it looks overly complex. 

To justify this I'm going to use this for the api key api too in a separate PR.

Let me know your thoughts.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/UserBoolQueryBuilder.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    private UserBoolQueryBuilder() {}
+
+    public static UserBoolQueryBuilder build(QueryBuilder queryBuilder) {
+        UserBoolQueryBuilder userQueryBuilder = new UserBoolQueryBuilder();
+        if (queryBuilder != null) {
+            QueryBuilder translaterdQueryBuilder = translateToUserQueryBuilder(queryBuilder);
+            userQueryBuilder.must(translaterdQueryBuilder);
+        }
+        userQueryBuilder.filter(QueryBuilders.termQuery("type", "user"));
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/UserBoolQueryBuilder.java:41]: For apikey this is called `doc_type`, just wanted to confirm that `type` for user is correct and expected here?

**Review by n1v0lg — COMMENTED** [review]
This is looking great! I've focused on the prod so far. I'll make a thorough pass over the tests tomorrow 👀 really nice work 💯

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/UserBoolQueryBuilder.java (hunk 3 above)
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/UserBoolQueryBuilder.java:41]: Thank you for doing that! Looks like this filter is acceptable based on the conversation and your comment here.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/QueryUserAction.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+package org.elasticsearch.xpack.core.security.action.user;
+
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.common.io.stream.Writeable;
+
+/**
+ * ActionType for retrieving Users
+ */
+public final class QueryUserAction extends ActionType<QueryUserResponse> {
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/QueryUserAction.java:16]: Nice! Didn't know about that one, will add it there instead.

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+                assertUser(testUsers.get(i), users.get(i));
+                // Only first element of array is used for sorting
+                assertThat(((List<String>) users.get(i).get("roles")).get(0), equalTo(((List<String>) users.get(i).get("_sort")).get(0)));
+            }
+        });
+
+        final String invalidFieldName = randomFrom("doc_type", "invalid", "password");
+        assertQueryError(400, "{\"sort\":[\"" + invalidFieldName + "\"]}");
+    }
+
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java:221]: Ah yes, that's a good thing to test. Will add a test for that.

**Review by n1v0lg — COMMENTED** [review]
Exciting stuff! A few test suggestions, and one more thing we gotta iron out:

Turns out that sorting on `full_name` and `email` does not work since they are `text` fields (see [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#fielddata-mapping-param)) -- I totally missed this when putting together the design doc, sorry! I think we should detect this and return a "friendly" error message. Note that _searching_ on them still works, it's just sorting that doesn't.

**Comment by jfreden** [comment]
Thanks for the review! 

> Turns out that sorting on full_name and email does not work since they are text fields (see [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#fielddata-mapping-param)) -- I totally missed this when putting together the design doc, sorry! I think we should detect this and return a "friendly" error message. Note that searching on them still works, it's just sorting that doesn't.

Yes, I noticed that too when testing. My thinking was that since the error handling is the same as for regular queries, it should be fine. Thinking about it some more though, the "regular error handling" is for indices that are user created, meaning the user setup the text mapping of the field, in this case we did so it's a little different. I'll add some better error handling for that.

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+            ),
+            Map.of(ESTestCase.randomAlphaOfLengthBetween(3, 8), ESTestCase.randomAlphaOfLengthBetween(3, 8)),
+            Map.of(),
+            null
+        );
+    }
+
+    private User createRandomUser() throws IOException {
+        return createUser(
+            randomValueOtherThanMany(reservedUsers::contains, () -> randomAlphaOfLengthBetween(3, 8)),
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java:355]: Good catch! Will fix.

**Comment by jfreden** [comment]
PR should be ready for another round of review @n1v0lg 
- Added a check and some tests for detecting sorting on `full_name` and `email`.
- Addressed your other great suggestions.

Thank you!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/QueryUserAction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/QueryUserAction.java]: Thanks!

**Review by n1v0lg — APPROVED** [review]
LGTM 🚀 🚀 🚀 

Only smaller comments, no need for re-review. Also, lets add a `test-serverless-update` label before merging. I don't expect this to break anything in serverless but easier to validate.

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+        User otherFieldsTestUser = createUser(
+            "batman-official-user",
+            new String[] { "bat-cave-admin" },
+            "Batman",
+            "batman@hotmail.com",
+            Map.of(),
+            true
+        );
+        String enabledTerm = "\"enabled\":true";
+        String fullNameTerm = "\"full_name\":\"Batman\"";
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java:132]: Yes, good catch!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexFieldNameTranslator.java (hunk 2 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexFieldNameTranslator.java:14]: @javanna I don't think @jfreden had immediate plans to work on this, so a PR is more than welcome! @jfreden is on PTO for the next two weeks
